debezium-core/src/main/java/io/debezium/transforms/UnwrapFromEnvelope.java

```diff
@@ -93,24 +131,45 @@ public R apply(final R record) {
             logger.warn("Expected Envelope for transformation, passing it unchanged");
             return record;
         }
-        final R newRecord = delegate.apply(record);
-        if (newRecord.value() == null && dropDeletes) {
-            logger.trace("Delete message {} requested to be dropped", record.key());
-            return null;
+        final R newRecord = afterDelegate.apply(record);
+        if (newRecord.value() == null) {
+            // Handling delete records
+            switch (handleDeletes) {
+                case "drop":
+                    logger.trace("Delete message {} requested to be dropped", record.key());
+                    return null;
+                case "rewrite":
+                    logger.trace("Delete message {} requested to be rewritten", record.key());
+                    final R oldRecord = beforeDelegate.apply(record);
+                    return removedDelegate.apply(oldRecord);
+                default:
+                    return newRecord;
+            }
+        } else {
+            // Handling insert and update records
+            switch (handleDeletes) {
```

Comment:
Is it necessary to handle deletes in both branches?

Reply:
Yep. Because we're adding __deleted = false field here.

Reply:
A question, dowe really need to add `__deleted` even for create and update records? Would it it be better to make the field optional and set it to`true` only for deletes?

Reply:
I think that even when this field will be optional we will need to add it with null value. So we still will be using InsertField for create / update records.
I will leave it this way. It's very easy to read.

Reply:
In my use case, I'm putting records on hadoop and then filtering them by __deleted = false condition.
I think it's more readable than __deleted is null.